### PR TITLE
[flutter_tools] avoid serving files outside of expected paths

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Switch to uri resolution over path combinations, and remove leading slashes to avoid looking up outside of cwd.

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart [hunk 1]
```diff
@@ -339,8 +342,9 @@ class WebAssetServer implements AssetReader {
     }
 
     if (!file.existsSync()) {
-      final String webPath = globals.fs.path.join(
-        globals.fs.currentDirectory.childDirectory('web').path, requestPath);
+      final Uri webPath = globals.fs.currentDirectory
+        .childDirectory('web')
+        .uri.resolve(requestPath);
```
Comment (ditman) [inline packages/flutter_tools/lib/src/build_runner/devfs_web.dart:347]: This is weird, I've been able to download the "pubspec.yaml" or some "json" files that are in the root of the app (siblings of `web`).

(Try: `localhost:port/pubspec.yaml`, or some `.iml`)

I haven't been able to escape outside of the root of the app!

**Review by ditman — COMMENTED** [review]
This is much better than before, but I can still access files that are siblings to the "web" directory:

* localhost:port/pubspec.yaml
* localhost:port/android/build.gradle

I guess we need to see "build" for assets and the rest are collateral damage?

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:347]: Its actually from above where we look for dart source. We need to resolve down from the project directory to handle requests like lib/main.dart for source maps, so this is working as expected I think

Comment (ditman) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:347]: Ahhh, makes sense. Thanks for the clarification!

**Review by ditman — APPROVED** [review]
LGTM!

**Comment by ditman** [comment]
(Apologies, I thought I had selected "Approve" last night!)

**Comment by jonahwilliams** [comment]
No worries, sorry for the trouble @ditman - but thanks for helping me fix it!

**Pull request merged**